plugins/inputs/kernel/psi.go

```diff
@@ -0,0 +1,74 @@
+//go:build linux
+
+package kernel
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/influxdata/telegraf"
+	"github.com/prometheus/procfs"
+)
+
+// Gather Psi metrics
+func (k *Kernel) gatherPressure(acc telegraf.Accumulator) error {
+	pressures, err := k.getPressureValues()
+	if err != nil {
+		return err
+	}
+	k.uploadPressure(pressures, acc)
+	return nil
+}
+
+// getPressureValues - Get the pressure values from /proc/pressure/*
+func (*Kernel) getPressureValues() (pressures map[string]procfs.PSIStats, err error) {
+	var fs procfs.FS
+	fs, err = procfs.NewDefaultFS()
+	if err != nil {
+		return nil, fmt.Errorf("procfs not available: %w", err)
+	}
+
+	pressures = make(map[string]procfs.PSIStats)
+	for _, resource := range []string{"cpu", "memory", "io"} {
+		pressures[resource], err = fs.PSIStatsForResource(resource)
+		if err != nil {
+			return nil, fmt.Errorf("failed to read %s pressure: %w", resource, err)
+		}
+	}
+	return pressures, nil
+}
+
+// uploadPressure Uploads all pressure value to corrosponding fields
+// NOTE: resource=cpu,type=full is omitted because it is always zero
+func (*Kernel) uploadPressure(pressures map[string]procfs.PSIStats, acc telegraf.Accumulator) {
+	for _, typ := range []string{"some", "full"} {
+		for _, resource := range []string{"cpu", "memory", "io"} {
+			if resource == "cpu" && typ == "full" {
+				continue
+			}
+
+			tags := map[string]string{
+				"resource": resource,
+				"type":     typ,
+			}
+
+			var stat *procfs.PSILine
+			switch typ {
+			case "some":
+				stat = pressures[resource].Some
+			case "full":
+				stat = pressures[resource].Full
+			}
+
+			now := time.Now()
+			acc.AddCounter("pressure", map[string]interface{}{
+				"total": stat.Total,
+			}, tags, now)
+			acc.AddGauge("pressure", map[string]interface{}{
+				"avg10":  stat.Avg10,
+				"avg60":  stat.Avg60,
+				"avg300": stat.Avg300,
+			}, tags, now)
+		}
+	}
+}
```

Comment:
Could we merge the two functions? Like
```suggestion
func (*Kernel) gatherPressure(acc telegraf.Accumulator) error {
	fs, err := procfs.NewDefaultFS()
	if err != nil {
		return fmt.Errorf("procfs not available: %w", err)
	}

	var metrics []telegraf.Metrics
	for _, resource := range []string{"cpu", "memory", "io"} {
		stats, err := fs.PSIStatsForResource(resource)
		if err != nil {
			return nil, fmt.Errorf("failed to read %s pressure: %w", resource, err)
		}

		// "Some" metrics total and average
		tags := map[string]string{
			"resource": resource,
			"type":     "some",
		}
		acc.AddGauge(
			"pressure",
			map[string]interface{}{
				"avg10":  stat.Some.Avg10,
				"avg60":  stat.Some.Avg60,
				"avg300": stat.Some.Avg300,
			},
			tags,
		)
		acc.AddCounter(
			"pressure",
			map[string]interface{}{
				"total": stats.Some.Total,
			},
			tags,
		)
		
		// "Full" metrics total and average
		tags = map[string]string{
			"resource": resource,
			"type":     "full",
		}
		acc.AddGauge(
			"pressure",
			map[string]interface{}{
				"avg10":  stat.Full.Avg10,
				"avg60":  stat.Full.Avg60,
				"avg300": stat.Full.Avg300,
			},
			tags,
		)
		// The total metrics is only conditionally added for "cpu" as this is all-empty and will be omitted.
		if resource != "cpu" {
			acc.AddCounter(
				"pressure",
				map[string]interface{}{
					"total": stats.Full.Total,
				},
				tags,
			)
		}
	}

	return nil
}
```

Reply:
Merging these functions sound like a good idea, but how could I implement unit tests? Since I'm relying on Prometheus to both read and parse `/proc/pressure/*`, there's pretty much nothing to test after merging collection and submission.

Reply:
It's probably a good idea to honor the `$HOST_PROC` environment variable like [gopsutil does](https://github.com/shirou/gopsutil?tab=readme-ov-file#usage) to allow running telegraf in a container which has the host's `/proc` mounted. This also has the great advantage to be able to fake the `/proc` dir for tests by placing the corresponding file-structure it in a `testdata` directory and set that environment variable accordingly...

Reply:
Oh I saw you already do have `psiDir` to specify where to find the stuff. So how about using `filepath.Dirname` on that path and then use [procfs's NewFS()](https://pkg.go.dev/github.com/prometheus/procfs#NewFS) instead of `procfs.NewDefaultFS()`?

Reply:
You reminded me of using [`procfs.NewFS`](https://pkg.go.dev/github.com/prometheus/procfs#NewFS) instead of `NewDefaultFS`, which is *the* solution we need.

Reply:
I think this is done now.